docs/architecture/adr-008-event-subscription.md

```diff
@@ -0,0 +1,519 @@
+
+![](assets/rpc-client-erd.png)
+
+### `Event`
+
+In terms of the subscription interface, this is ultimately what the end user is
+most interested in obtaining. The `Event` type's structure is dictated by the
+Tendermint RPC:
+
+```rust
+pub struct Event {
+    /// The query that produced the event.
+    pub query: String,
+    /// The data associated with the event (determines its `EventType`).
+    pub data: EventData,
+    /// Event type and attributes map.
+    pub events: Option<HashMap<String, Vec<String>>>,
+}
+
+pub enum EventData {
+    NewBlock {
+        block: Option<Block>,
+        result_begin_block: Option<BeginBlock>,
+        result_end_block: Option<EndBlock>,
+    },
+    Tx {
+        tx_result: TxResult,
+    },
+    // ...
+}
+```
+
+### `Subscription`
+
+A `Subscription` here is envisaged as an entity that implements the
+[Stream][futures-stream] trait, allowing its owner to asynchronously iterate
+through all of the relevant incoming events. Use of such a subscription should
+be as simple as:
+
+```rust
+while let Some(result_event) = subscription.next().await {
+   match result_event {
+      Ok(event) => { /* handle event */ },
+      Err(e) => { /* terminate subscription and report error */ },
+   }
+}
+```
+
+For efficient routing of events to `Subscription`s, each `Subscription` should
+have some kind of unique identifier associated with it (a `SubscriptionId`).
+Each `Subscription` relates only to a single [`Query`](#query). Therefore, its
+publicly accessible fields may resemble the following:
+
+```rust
+pub struct Subscription {
+    pub id: SubscriptionId,
+    pub query: Query,
+    // ... other fields to help facilitate inter-task comms ...
+}
+```
+
+`Subscription`s are created by a client - described in the following sub-section.
+
+### Client Model
+
+Users of the Tendermint RPC library may or may not want access to subscription
+functionality. Since such functionality comes with additional overhead in terms
+of resource usage and asynchronous task management, it would be optimal to
+provide two separate client traits: one that only implements non-subscription
+functionality, and one that only implements subscription functionality (where
+clients could either implement one or both traits).
+
+The interfaces of the two types of clients are envisaged as follows.
+
+#### `Client`
+
+This type of client would allow for interaction with all RPC endpoints except
+those pertaining to subscription management. In our current implementation, this
+client would only interact via the HTTP RPC endpoints (the `HttpClient` in the
+entity diagram above).
+
+**Note**: All `async` traits are facilitated by the use of [async-trait].
+
+```rust
+pub type Result<R> = std::result::Result<R, Error>;
+
+#[async_trait]
+pub trait Client {
+    /// `/abci_info`: get information about the ABCI application.
+    async fn abci_info(&self) -> Result<abci_info::AbciInfo>;
```

Comment:
Not sure how valuable it is to list more than a few of these here since the tendermint RPC endpoints are already well defined and the methods all follow a common pattern. So just need an indication that there's a method for every RPC endpoint, some representative examples, any unique aspectss, and then the perform method

Reply:
Good point - I'll condense this 👍